src/collector.rs

```diff
@@ -1,63 +1,282 @@
+
+        let input_content = InputContent {
+            input: Input::Stdin,
+            file_type: file_type_hint.unwrap_or_default(),
+            content,
+        };
+
+        Ok(input_content)
+    }
+
+    fn string_content(s: &str, file_type_hint: Option<FileType>) -> Result<InputContent> {
+        Ok(InputContent::from_string(
+            s,
+            file_type_hint.unwrap_or_default(),
+        ))
     }
 }
 
-/// Fetch all unique links from a vector of inputs
+impl ToString for Input {
+    fn to_string(&self) -> String {
+        match self {
+            Self::RemoteUrl(url) => url.to_string(),
+            Self::FsGlob { pattern, .. } => pattern.clone(),
+            Self::FsPath(p) => p.to_str().unwrap_or_default().to_owned(),
+            Self::Stdin => STDIN.to_owned(),
+            Self::String(s) => s.clone(),
+        }
+    }
+}
+
+/// Fetch all unique links from a slice of inputs
 /// All relative URLs get prefixed with `base_url` if given.
 pub(crate) async fn collect_links(
-    inputs: Vec<String>,
+    inputs: &[Input],
     base_url: Option<String>,
+    skip_missing_inputs: bool,
 ) -> Result<HashSet<Uri>> {
     let base_url = match base_url {
         Some(url) => Some(Url::parse(&url)?),
         _ => None,
     };
 
-    let mut links = HashSet::new();
+    let (contents_tx, mut contents_rx) = tokio::sync::mpsc::channel(32);
 
-    for input in inputs {
-        match Url::parse(&input) {
-            Ok(url) => {
-                let path = String::from(url.path());
-                let res = reqwest::get(url).await?;
-                let content = res.text().await?;
+    // extract input contents
+    for input in inputs.iter().cloned() {
+        let mut sender = contents_tx.clone();
 
-                links.extend(extract_links(
-                    resolve_file_type_by_path(path),
-                    &content,
-                    base_url.clone(),
-                ));
-            }
-            Err(_) => {
-                // Assume we got a single file or a glob on our hands
-                for entry in glob(&input)? {
-                    match entry {
-                        Ok(path) => {
-                            let content = fs::read_to_string(&path)?;
-                            links.extend(extract_links(
-                                resolve_file_type_by_path(&path),
-                                &content,
-                                base_url.clone(),
-                            ));
-                        }
-                        Err(e) => println!("Error handling file pattern {}: {:?}", input, e),
-                    }
-                }
-            }
-        };
+        tokio::spawn(async move {
+            let contents = input.get_contents(None, skip_missing_inputs).await;
+            sender.send(contents).await
+        });
+    }
+
+    // receiver will get None once all tasks are done
+    drop(contents_tx);
```

Comment:
Interesting. I forgot that for the client_pool channel. Is that always needed? If so I will add it.

Reply:
According to the [tokio docs](https://docs.rs/tokio/0.2.23/tokio/sync/index.html#mpsc-channel):

> 
    // The `rx` half of the channel returns `None` once **all** `tx` clones
    // drop. To ensure `None` is returned, drop the handle owned by the
    // current task. If this `tx` handle is not dropped, there will always
    // be a single outstanding `tx` handle.

Reply:
Ouch then I'll add the explicit drop. 👍